### ISSUE TITLE
variablesConverter is not working correctly.

### ISSUE THREAD

**Issue by jeong-sik** [zth/rescript-relay#287]
![image](https://user-images.githubusercontent.com/45959991/149865929-9fe65707-493e-46cd-95e0-023abdf3eab7.png)
The expected value is

searchInput: {
  pesticideIds: [3518],
  take: 1,
  skip: 0
}

It probably goes through the`utils.traverser` function and the value is wrong.

**Comment by jeong-sik** [comment]
└─ rescript-relay@1.0.0-beta.7

**Comment by zth** [comment]
Thank you for the report! What does your Relay config look like, do you have any custom scalars defined?

**Comment by jeong-sik** [comment]
> 
Here! Thank you!
```js
// relay.config.js
module.exports = {
  src: "./src", // Path to the folder containing your Reason files
  schema: "./schema.graphql", // Path to the schema.graphql you've exported from your API. Don't know what this is? It's a saved introspection of what your schema looks like. You can run `npx get-graphql-schema http://path/to/my/graphql/server > schema.graphql` in your root to generate it
  artifactDirectory: "./src/__generated__", // The directory where all generated files will be emitted

  // You can add type definitions for custom scalars here.
  // Whenever a custom scalar is encountered, the type emitted will correspond to the definition defined here. You can then deal with the type as needed when accessing the data.
  customScalars: {
    DateTime: "string",
    DateTimeClj: "string",
    DateClj: "string"
  }
}
```

```graphql
input PesticideListSearchInput {
  companyName: [String!]
  pesticideIds: [Int!]
  skip: Int!
  take: Int!
}
```

**Comment by zth** [comment]
Thank you! Could you also paste the full `variablesConverter`, the variables definition of the mutation you're doing (`mutation Blabla(<this stuff>) {`), as well as any other input object from the schema that appears in there?

**Comment by jeong-sik** [comment]
All the query code looks like this:

```res
module PesticideGraphQL_PesticideCommon_Fragment = %relay(`
  fragment PesticideGraphQL_PesticideCommon_Fragment on Pesticide {
     _id: id
     madeIn
     productName
     contentForm
     ratio
     toxicity
     purpose
     aquaticToxicity
     companyName
     registrationNumber
  }
`)

module PesticideGraphQL_PesticideImage_Fragment = %relay(`
  fragment PesticideGraphQL_PesticideImage_Fragment on PesticideImage {
     type_ : type
     order
     originName
     url
  }
`)

module PesticideGraphQL_PesticidePest_Fragment = %relay(`
  fragment PesticideGraphQL_PesticidePest_Fragment on PesticidePest {
     code1Name
     pestName
     whenToUse
     howToUse
     dilutionRatio
     safeUseDate
     safeUseCount
  }
`)

module PesticideGraphQL_PesticideIngredient_Fragment = %relay(`
  fragment PesticideGraphQL_PesticideIngredient_Fragment on PesticideIngredient {
     korName
     engName
     ratio
     mechanism
  }
`)

module PesticideGraphQL_Detail_ById_Query = %relay(`
  query PesticideGraphQL_Detail_ById_Query($searchInput: PesticideListSearchInput, $sameIngredientSearchInput: GlCropNameInput) {
    pesticides (input: $searchInput, common: {}) {
     ...PesticideGraphQL_PesticideCommon_Fragment
     pest {
       ...PesticideGraphQL_PesticidePest_Fragment
     }
     ingredient {
       ...PesticideGraphQL_PesticideIngredient_Fragment
     }
     image {
       ...PesticideGraphQL_PesticideImage_Fragment
     }
     sameIngredient(input: $sameIngredientSearchInput) {
        ...PesticideGraphQL_PesticideCommon_Fragment
        image {
          ...PesticideGraphQL_PesticideImage_Fragment
        }
      }
    }
  }
`)
```

```graphql
input GlCropNameInput {
  glCrop: [String!]
}
```

**Comment by zth** [comment]
Thank you! I found the issue, new beta with the fix coming tonight. Thanks for helping me track this down, this was a nasty little bug in my Rust rewrite that I've been chasing for a good while now.

**Comment by mattdamon108** [comment]
Can I ask what you mean `Rust rewrite`?

**Comment by zth** [comment]
> Can I ask what you mean `Rust rewrite`?

Of course! It's detailed in this blog post here: https://rescript-relay-documentation.vercel.app/blog/2021/01/03/rescript-relay-1.0.0-beta

But basically, for `1.0.0`, the entire type generation has been rewritten from scratch, from a combination of TS/ReasonML to Rust. It has also been integrated fully into a forked version of the new Relay compiler (which is written in Rust, therefore the typegen for RescriptRelay is now also written in Rust).

That rewrite mostly went really well, but there are a lot of subtleties and quite a few edge cases, this particular one being among them.

**Comment by mattdamon108** [comment]
Super interesting. I have some question about it. Currently rescript-relay-ppx works as ppx in rescript codes. Can you explain how to bridge between the rescript ppx in rescript code and type writter which is written in Rust.

In my guess, does your new type writter is parsing the graphql schema and write rescript type files directly not through ppx?

**Comment by zth** [comment]
Exactly! So `rescript-relay` has never actually emitted any types via PPX, like `graphql-ppx` does. The Relay compiler produces types in actual generated files, and what I've built into the Relay compiler is the ability to emit ReScript types and assets, where as the core compiler only supports Flow and TypeScript by default.

The PPX really only does two, fairly minor things (but very crucial to the DX):
1. It ties together the module defining a query/fragment/sub/mutation with the file the Relay compiler generates from it. This is what makes it feel like the PPX is generating the types, where as the types are actually in a generated file on disk.
2. It ensures that "illegal" functions can't be used with Relay. For example, in order to be able to do `Fragment.useRefetchable`, there must be a `@refetchable` directive in the fragment. So the PPX checks for that, and only exposes `useRefetchable` if that directive exists. And a bunch of other similar things, only exposing certain functions when the requirements are met. This is important because it means that it's impossible to use things "wrong".

Hope that makes sense!

**Comment by zth** [comment]
1. You write a GraphQL definition in `%relay()`
2. The Relay compiler finds that, extracts the GraphQL, and emits a generated file with ReScript types and assets
3. The PPX, in a very simplified sense, replaces `%relay()` with  an alias to the generated file.

**Comment by mattdamon108** [comment]
It does totally make sense. I was curious about how to parse and generate types for the fragments, queries, and mutation written in recript codes without ppx. I got your point that ppx still will work for it.

So, the type writter in Rust will generate the basic and fundermental types on disk from the graphql schema file and ppx will generate types after parsing the fragments, queries, and mutations on rescript code.

Is it right?

**Comment by mattdamon108** [comment]
Oh I misunderstood. It seems the ppx is only for making alias to link the types on disk written by Rust writter.

**Comment by mattdamon108** [comment]
I think it is very clever design! I’m very interesting to see how the rescript types will be generated in Rust code.

I’m looking forward to seeing your work soon!

**Comment by zth** [comment]
@mattdamon108 Here's the Rust compiler fork including my changes: https://github.com/facebook/relay/compare/main...zth:rescript-relay

It's obviously pretty involved with the core workings of the Relay compiler. But, in a nutshell, the Relay compiler gives me an AST designed to print Flow/TS. I take that AST and piece it together into a representation that's what I need to print ReScript types and assets. I then print the ReScript types and assets using regular string concatenation.

**Comment by mattdamon108** [comment]
Super, indeed!
I’ll also look at it.

**Comment by zth** [comment]
@jeong-sik this should be fixed in `1.0.0-beta.8` that was just released 🎉 Closing this for now.
